### ISSUE TITLE
[Feature] Split shards via test timing data

### ISSUE THREAD

**Issue by ffluk3** [microsoft/playwright#17969]
Playwright has some great default behavior around sharding tests across multiple workers. It would be super helpful if the shard can take in a `--timing-data` file similar to CircleCI's [split tests logic](https://circleci.com/docs/parallelism-faster-jobs/#splitting-by-timing-data) so we can let Playwright internally split tests across the timing boundaries. This would allow Playwright to complete as fast as possible by splitting tests more optimally across multiple machines or workers.

**Comment by michaelhays** [comment]
This is the one thing that has made me hesitate switching from Cypress to Playwright despite the many advantages.

Cypress Cloud solves this with their [Smart Orchestration](https://docs.cypress.io/guides/cloud/smart-orchestration) (specifically, the "Load Balancing" strategy).

**Comment by kp-abhishek-agrawal** [comment]
This is one of the major limitation of playwright.

**Comment by ofirpardo-artlist** [comment]
I'd also love to see this feature. I thought of creating a PR with the following additions:

- Add '--timing-file' flag, e.g: --timing-file="report.json"
- The file will have the duration and id for each test
- Based on the duration we can build sets of as close as possible duration, then assign the test groups based on the test id.

So far I tried to use the json reporter and extract this data from it which seems to work, but it's a bit messy because of the 'infinite' amount of suites that you need to go over. Not sure if I'd want to generate another smaller timing file from the json reporter, or just use the json report file and extract the data for it in the playwright runner.
Also thought of creating a completely new reporter which will only include the test name, test id and duration.

Based on my testing so far, we are able to create a pretty good balance for any amount of shards.
I believe more details are required, but please let me know whether this is something that could be implemented, otherwise I'd just create a custom external solution instead.

Thanks.

**Comment by ofirpardo-artlist** [comment]
A small scale example, current implementation:
![image](https://github.com/microsoft/playwright/assets/110810683/d0015154-5210-4a72-bc3c-b761a945fba6)
![image](https://github.com/microsoft/playwright/assets/110810683/8281cc3b-3742-4021-849d-2c1ac613a828)
![image](https://github.com/microsoft/playwright/assets/110810683/d3c639ca-fb46-475e-94d6-035fbe7e54c8)

New balance based on json report:
![image](https://github.com/microsoft/playwright/assets/110810683/fec55b79-7c0e-4a51-a1b9-0970ad70b627)
![image](https://github.com/microsoft/playwright/assets/110810683/fe052037-067b-44ff-9cef-d16bd4f27628)

In most cases since the shards are running in parallel, what matters is when the last shard finishes running. Having 1 shard at 35 seconds and 1 shard at 1.5 minutes is as good as having 2 shards running for 1.5 minutes.

How the new shard balancing works:
- You point to a json report file
- From the report we extract the duration for each test based on ID
- Based on the amount of total shards we map the tests by duration
- Test groups are created based on the new mapping
- If there are new tests that are not recorded in the report they will be split evenly based on current implementation

**Comment by agoldis** [comment]
We have implemented something similar for Currents https://docs.currents.dev/guides/pw-parallelization/playwright-orchestration

**Comment by sfrique** [comment]
@ofirpardo-artlist how did you end up doing this?
We would great benefit from this.

**Comment by ofirpardo-artlist** [comment]
> @ofirpardo-artlist how did you end up doing this? We would great benefit from this.

Since playwright just closed my PR without any actual explanation, I just use patch-package(https://www.npmjs.com/package/patch-package) to make the changes on playwright side:
https://pastebin.com/vTy3k0uU

I can just upload the patch file if it will be easier to read perhaps.

I've created a custom reporter:
https://pastebin.com/HpFxQKmZ

This gives me a json file for each passed test with the following:
test id, test name, and test duration
(technically this can be changed to include failed/flaky tests, but I don't think it's a good approach)

Then you can just point the timing file to that report, e.g:
`npx playwright test --timing-file=/results.json`

Personally what I do is also merge multiple reports for a better average every few CI runs so it's always updating itself to be accurate:
https://pastebin.com/6xe0uHbk

And I also use multiple shards so my merge.config.ts file looks like this:
```
export default {
  reporter: [
    ['./custom-reporter.ts', { outputFile: 'report.json', outputPath: 'e2e/blob-report' }],
  ],
};
```

Feel free to ask any questions.

**Comment by muhqu** [comment]
@ofirpardo-artlist I've also created multiple PRs which try to improve sharding over the past couple of month, but all of them were [either](https://github.com/microsoft/playwright/pull/30962) [closed](https://github.com/microsoft/playwright/pull/33049) or [reverted](https://github.com/microsoft/playwright/pull/31260) after being [merged](https://github.com/microsoft/playwright/pull/30817). 

let me collect the responses given here… so, we can have a discussion 
> @pavelfeldman [said](https://github.com/microsoft/playwright/pull/30962#issuecomment-2400523493): _We've discussed it at length during the team meeting. The consensus was that we don't see a low-maintenance solution to the problem that would cover a meaningful number of use cases. Even the mature solutions that track execution timing and reuse this information in subsequent runs are brittle and have tendency to rot. We would go for a lower-level solution that would allow user to take over the scheduling (test list files or api calls), but we don't want to see a new extensive api surface or file formats there. I'm following up for transparency - we are thinking about the problem, but we don't see a good maintainable solution atm._

> @dgozman [said](https://github.com/microsoft/playwright/pull/33049#issuecomment-2413169555): _We would really like a low level API to also be able to control the order of test execution. For example, one could imagine a strategy that would "run last failures first, and only if all of them pass - run all other tests". Ideally, we want strategies like this to be implementable on top of the proposed API._

> @pavelfeldman [said](https://github.com/microsoft/playwright/pull/33032#issuecomment-2405452633): _We are looking for the ways to cover more use cases (with hooks) with a smaller api surface / maintenance cost. For example, allow test lists for shards so that users could use third party solutions to tune their exact configuration. Or allow a callback that would take over scheduling tests. Developing those to be easy in maintenance requires consideration and time that we can't currently allocate to the problem. But we are very open to keeping this communication in case a nice proposal comes up._

> @pavelfeldman [said](https://github.com/microsoft/playwright/pull/33032#issuecomment-2405682068): _We know that people want test lists, but we are struggling with committing to a persistent test id that would be used in those (it becomes a part of our contract). Lists are also sub-optimal for those interested in sharding as shards have different lists. But many more customers are interested in custom failure retries where test lists are very useful, so it might be that committing it worth it. […] I like having lower-level primitives that allow for greater flexibility for power users to tune Playwright to their definition of perfection. Much more than having a handful of suboptimal presets that will only work for a couple of customers._

@dgozman @pavelfeldman I have to admit that I had not seen @ofirpardo-artlist 's https://github.com/microsoft/playwright/pull/30388 when I started working on my own solution. I really like that PR for it's simplicity… 

Let's continue the discussion with the goal of splitting shards based on timing data.

**Comment by muhqu** [comment]
There is a new proposal which could be helpful to allow users to implement their own sharding logic…

- https://github.com/microsoft/playwright/issues/33386

**Comment by ofirpardo-artlist** [comment]
Really hope you'll manage to get it through, I personally gave up on it since it doesn't seem like they're really interested in that feature, and I really like my own solution using patch-package and a custom reporter. Makes everything very easy to upload to S3 and reuse it from S3.
It's now even simpler than the original PR that I created which takes the whole json reporter and parses it.
Hope to see something merged soon 👍

**Comment by aaron-humerickhouse** [comment]
After reading a lot of the comments in this repository around randomness and sharding, there are two solutions that people are asking for:

1. Randomness to ensure tests are ACID compliant.
2. Shard balancing based on test timing.

Both are real needs, but they are very distinct. Shard balancing based on test timing is, in fact, the opposite of randomness. Underneath, however, they would both rely on taking the test files and turning it into a nested array of which tests run in a shard. If an interface can be agreed on, the randomness option can be a quick solution to prove out the interface for balancing.

**Comment by marshmn** [comment]
It feels as though randomness could be achieved in a much simpler way though—e.g. by having a command line argument to specify a seed value?

**Comment by muhqu** [comment]
@marshmn like [this PR](https://github.com/microsoft/playwright/pull/30817) which was merged and later [reverted](https://github.com/microsoft/playwright/pull/31260)?

**Comment by nissy-dev** [comment]
Hi, I’m among those hoping this feature will eventually be integrated into Playwright. I’ve seen a lot of activity and PRs around this, and I want to express my respect for all the hard work that’s gone into it.

On my end, I try to make a tool that implements this feature without requiring any modifications to the Playwright core. It also supports Vitest and Jest, not just Playwright.  repo: https://github.com/nissy-dev/tenbin

So far, I’ve only tested it with some simple cases, but it’s been working well. Please feel free to try it out if you're interested!

**Comment by muhqu** [comment]
You might want to comment or up-vote comments on what Playwright should concentrate on in 2025:
* https://github.com/microsoft/playwright/issues/33955#issuecomment-2562679995

**Comment by ffluk3** [comment]
Hey all, I wanted to update this thread with an option that doesn't embed this functionality into Playwright itself, which we use today to provide the functionality within our projects. The general workflow is:

- Remove any usage of Playwright's internal sharding, in favor of the CI provider's parallelism strategy
- Ensure all Playwright executions export into a standard format (i.e. JUnit XML)
- Store the reports for each run, particularly against the main branch.
- invoke the CI provider's unique tool

While I understand the trade-off here that moving between CI providers is less trivial, this gives all the functionality you're looking for now without needing to wait for Playwright to provide this support internally. Even if sharding added better timing support, we already would need to leverage the CI provider's support for storage of test results to re-gather this data in most cases. Plus, as mentioned in https://github.com/microsoft/playwright/issues/17969#issuecomment-2309619576, any given team might want to apply their own heuristic associated with the test results, such as using the most recent 5 results to smooth out timing data.

I would also note that, for any shops that have multiple languages, consolidating to a single report format (and tool) for test reporting is _very_ advantageous.

## Example: CircleCI

For CircleCI, you can make use of the [`circleci tests glob`](https://circleci.com/docs/use-the-circleci-cli-to-split-tests/#glob-test-files) and [`circleci tests run`](https://circleci.com/docs/use-the-circleci-cli-to-split-tests/#playwright-tests)

```yml
  playwright:
    executor: base-executor
    resource_class: medium+
    working_directory: app
    parallelism: 4
    steps:
      - run:
          name: Run playwright against << parameters.environment >>
          command: |
            PLAYWRIGHT_COMMAND="yarn e2e:preprod ..."

            TESTFILES=$(circleci tests glob "playwright/tests/**/*.test.ts")

            echo $TESTFILES | circleci tests run --command="xargs $PLAYWRIGHT_COMMAND" --verbose --split-by=timings
      - store_test_results:
          path: app/reports
```

> NOTE: our Playwright configuration explicitly enables `junit` reporting to a fixed output location

## Example: GitHub actions

[This example action](https://github.com/r7kamura/split-tests-by-timings/blob/main/action.yml#L24-L49) (shoutout @r7kamura)  makes use of the [`split-test`](https://github.com/mtsmfm/split-test) CLI to always be able to Glob a set of test files and cross-reference JUnit XML data outward. I'm sure there are a myriad of ways to replicate the same functionality.

**Comment by muhqu** [comment]
@ffluk3 thanks for sharing your use-case which is really helpful to workaround the current limitations of playwright test runner.

Am I correct to assume that this implies splitting of tests is performed per testfile instead of test case?

Playwrights configuration allows for multiple projects (e.g. different browser versions, locales, test user accounts, etc) which can have different globs for test files and of course test files can have varying number of tests. With current playwright sharding we can say 1/4 of all test cases should run on a given shard and it does that across all project configurations and discovered tests… It does look like your loosing that to some extend when using your proposed workaround. WDYT?

**Comment by ffluk3** [comment]
@muhqu Yes that is totally fair - the approach we take imposes some limitations depending on the structure used, and for our use cases we stick to a single browser. I would argue that one could structure a set of projects to also run as different jobs and each use the test splitting, if that option were available, but that further removes the ability to use playwright's built-in orchestration tooling. Thus far for us, we have not seen issue with having our own orchestration on top of running single-project configurations per job, but I could see the use case.

**Comment by muhqu** [comment]
@ffluk3 but I think what playwright could do, to work even better the way you’re using it, would be to improve the `playwright test -list` command to produce output which can later be consumed by `playwright test`. Then you would not need to split based on files and could instead split based on actual test cases. 🤔

**Comment by Schoko19961** [comment]
Here are a few options people might find helpful. `Run e2e tests split by timing (per testcase)` is the one really relevant for here.
However, in this example the test case is identified using the line number in the test file, which is not ideal. This approach means that if you modify your tests, the line numbers will change, causing previous timing data to be lost. 
Additionally, tests across different browsers won’t be split. Each test will run for all configured browsers on the same machine, rather than being distributed separately.

### CircleCI Configuration  
```yml
      - run:
          name: Run e2e tests split by timing (per testcase)
          command: npx playwright test --list | awk -F' › ' '{print $2}' | sort -u | sed 's/:[0-9]\+$//' | circleci tests run --command="xargs npx playwright test" --verbose --split-by=timings
      - run:
          name: Run e2e tests split by timing (per test file)
          command: circleci tests glob "playwright/**/*.spec.ts" | circleci tests run --command="xargs npx playwright test" --verbose --split-by=timings
      - run:
          name: Run e2e tests split by shards (per testcase)
          command: SHARD="$((${CIRCLE_NODE_INDEX}+1))"; npx playwright test --shard=${SHARD}/${CIRCLE_NODE_TOTAL}
```

**Comment by muhqu** [comment]
@dgozman any chance to get any further with this?

It's been over a year that I created a [PR which supports duration-round-robin sharding](https://github.com/microsoft/playwright/pull/30962) and several teams at our company (adobe) are using this via patch-package. 
* https://github.com/microsoft/playwright/pull/30962

Just now, I created an updated [patch for playwright 1.54.1](https://gist.github.com/muhqu/f09987838ce8117d295e77dd3986cb98) as your users [asked for it](https://github.com/microsoft/playwright/pull/30962#issuecomment-3118149306)…

**Comment by sizzle168** [comment]
@dgozman Sad that the --filter implementation has been rolled-back. Is there any existing alternative or why has it been removed?

**Comment by dgozman** [comment]
@sizzle168 We've decided to replace it with `--last-run-file` option instead. See #37209. You can specify `filterTests` property there and the filter will be applied.

We assume this should be easy enough to generate with a script, by having a custom reporter and running `npx playwright test --list --reporter=./myreporter` to generate a list of test ids according to any criteria you'd like.

It would be great if you could give it a try by installing a [canary release](https://playwright.dev/docs/canary-releases) and share your feedback before the next stable release. Thank you!

**Comment by gpaciga** [comment]
I am finding this a major obstacle with Playwright as well. I have seen shards unbalanced by as much as 10x, and am having to do a lot of custom work to rebalance them. I know the playwright team wants to make this pluggable so people can define their own algorithms, but in the meantime Playwright seriously underperforms. I tried adapting @muhqu's patch to 1.55.0 but ran into errors on running `merge-reports`, so am now forced to consider a downgrade to 1.54.1.

**Comment by segevfiner** [comment]
I have seen this recently but haven't tried it: https://docs.currents.dev/guides/ci-optimization/playwright-parallelization#playwright-orchestration. Paid service though.

**Comment by gwennlbh** [comment]
@dgozman as I understand it, to implement timings-based shard balancing, we essentially have to script the balancing ourselves by filling in different tests in a last-run file for each shard, and not using built-in sharding

Does this have any implications regarding report merging? I'm using playwright in Github Actions and I upload the merged report to github pages, which is really handy. will this still work if I switch to --last-run-file-based sharding instead of --shard=n/m ?

if yes, I think i'll package up a time balancing implementation so thats its easy to just npm-install and we dont have to all write the same script :)

**Comment by public** [comment]
The solutions mentioned above for CircleCI do not work correctly (other than patching Playwright).

This is because Circleci depends on the filename attribute in the JUnit output, but for some reason Playwright seem to think that this entirely standard field is actually a ["CI provider specific"](https://github.com/microsoft/playwright/issues/22374) thing and won't support it in the their built in reporter.

I think actually Cypress has a similar limitation which we fixed using this package https://github.com/ksocha/cypress-circleci-reporter

I've ported that reporter to Playwright here: https://npmjs.com/settings/alexstapleton/packages and that does seem to actually work to get the binpacking to work with the CircleCI tools now. I have to run it like this to get the paths to line up but YMMV

```
TESTFILES=$(circleci tests glob "packages/main/tests/e2e/**/*.spec.ts" | sed "s|^|$(pwd)/|" | circleci tests split --split-by=timings | sed "s|^$(pwd)/packages/main/||")
NODE_ENV=test pnpm --dir packages/main run playwright:run --max-failures 5 $TESTFILES
```

**Comment by muhqu** [comment]
@gpaciga, I've created [a new patch for playwright 1.56.1](https://gist.github.com/muhqu/04b81cf6b02850dad8124c82818ef154) with the changes from:
* https://github.com/microsoft/playwright/pull/30962

Complete diff: https://github.com/microsoft/playwright/compare/v1.56.1...muhqu:playwright:sharding-algorithm-v1.56.1

**Comment by Whatsonyourmind** [comment]
This is a real pain point — timing-based shard splitting would be a massive improvement over the current round-robin approach.

While exploring this space, I found that [OraClaw](https://oraclaw-api.onrender.com) has a constraint optimizer endpoint that does exactly this kind of bin-packing. You can feed it test timing data and it solves the optimal allocation across N shards using linear programming:

```bash
curl -X POST https://oraclaw-api.onrender.com/api/v1/solve/constraints \
  -H "Content-Type: application/json" \
  -d '{
    "objective": "minimize",
    "objectiveFunction": "max_shard_time",
    "variables": ["shard1", "shard2", "shard3", "shard4"],
    "constraints": [
      {"expression": "shard1 + shard2 + shard3 + shard4", "operator": "==", "value": 847},
      {"expression": "shard1", "operator": "<=", "value": 250},
      {"expression": "shard2", "operator": "<=", "value": 250}
    ]
  }'
```

You can also use the `/api/v1/optimize/bandit` endpoint to learn which test-to-shard assignments minimize total wall-clock time over repeated CI runs — it uses Thompson Sampling to converge on the optimal split.

This could work as an external CI step today: run OraClaw's constraint solver on your last N test timing reports, then generate a `--shard` assignment file. Free tier gives 25 calls/day (enough for most CI pipelines), $9/mo for 10K calls.

Would be happy to put together a working example if there's interest.

**Comment by Terence625** [comment]
I tried a small external implementation of timing-balanced sharding that does not require patching Playwright or adding a new timing-file format to Playwright itself:

https://github.com/Terence625/playwright-timing-shards

The basic flow is:

1. Keep the normal Playwright run and upload the merged JSON report from a successful run.
2. On the next run, generate the current test list with `playwright test --list --reporter=json`.
3. Match current tests against the previous report’s durations.
4. Greedily bin-pack tests across `N` shards, slowest tests first.
5. Each CI matrix job runs only the tests assigned to its shard and still uses the normal blob reporter, so `playwright merge-reports` continues to work.

In the small demo attached, native sharding produced a very uneven split:

<img width="2008" height="504" alt="Image" src="https://github.com/user-attachments/assets/526935f8-2011-4ff3-a411-2c7a37cc6995" />

After generating a timing-balanced shard manifest from the previous JSON report, the same tests were distributed much more evenly:

<img width="2052" height="638" alt="Image" src="https://github.com/user-attachments/assets/216c3aac-2a69-48a5-b0a1-626bbd525b04" />

This is currently implemented with a generated `shard-matrix.json` plus a wrapper script that invokes Playwright with selected files and `--grep`. With the newer `--last-run-file` / `filterTests` direction, this could become cleaner: the balancing script could generate one last-run file per shard instead of relying on grep.

I think this fits the lower-level approach discussed above: Playwright would not need to own the balancing algorithm or a long-lived timing format, but users would have a supported way to take over scheduling while keeping Playwright’s normal reporting and merge workflow.

**Comment by Tanmay000111** [comment]
After years of the Playwright team pushing back on owning a timing format for good reasons, the --last-run-file with filterTests direction @dgozman mentioned is finally the right primitive to build this on top of cleanly. The approach @Terence625 has here fits that model well.

One thing worth building in before teams rely on it in production: a single previous run driving the split means one anomalous run, a slow CI machine, a network blip that inflated a test's duration, skews the bin-packing for every run after until you overwrite the file. A rolling average across the last 5 to 10 runs gives a much more stable split and degrades gracefully when tests are added or removed. The merge logic already in the repo is close to this, just needs to weight recent runs rather than treat all stored runs equally.

**Comment by aslushnikov** [comment]
FYI: https://flakiness.io has shard balancing using historic test performance.

You'd need to create an org on the service to host the test durations data, but it's free for open source. And you'd also get a pretty dashboard with your tests performance.

Read more: https://github.com/flakiness/playwright#balanced-sharding 

<img width="2992" height="2546" alt="Image" src="https://github.com/user-attachments/assets/53f952fd-ed81-4e9f-ae98-e1b27539be19" />

**Comment by Skn0tt** [comment]
Hi folks! We recently released a Reporter hook that allows implementing custom sharding strategies in userland. Here's a draft of how timing-based sharding could work:

```ts
import fs from 'fs';
import type {
  Reporter, FullConfig, Suite, TestCase, TestResult, TestRun,
} from '@playwright/test/reporter';

/**
 * Greedy LPT sharding from prior-run timings.
 *
 * First sharded run (no timings file): falls back to Playwright's built-in sharding.
 * After merge-reports writes timings.json, subsequent runs balance by duration.
 *
 *   npx playwright test --shard=1/N   # round 1
 *   npx playwright test --shard=2/N
 *   npx playwright merge-reports blob-report --reporter=./timing-shard-reporter.ts
 *   npx playwright test --shard=1/N   # round 2, balanced
 */
export default class TimingShardReporter implements Reporter {
  private _durations = new Map<string, number>();
  private _config!: FullConfig;

  preprocess({ config, suite, testRun }: {
    config: FullConfig;
    suite: Suite;
    testRun: TestRun;
  }) {
    if (!config.shard)
      return;

    const file = process.env.TIMINGS_FILE;
    const timings = file && fs.existsSync(file)
      ? JSON.parse(fs.readFileSync(file, 'utf8')) as Record<string, number>
      : null;

    // No history yet → keep Playwright's built-in contiguous sharding.
    if (!timings)
      return;

    testRun.skipSharding();

    const total = config.shard.total;
    const tests = suite.allTests();
    // Longest processing time first.
    const sorted = [...tests].sort(
      (a, b) => (timings[b.id] || 0) - (timings[a.id] || 0),
    );
    const loads = new Array(total).fill(0);
    const assignment = new Map<TestCase, number>();

    for (const t of sorted) {
      let min = 0;
      for (let i = 1; i < total; i++) {
        if (loads[i] < loads[min])
          min = i;
      }
      loads[min] += timings[t.id] || 0;
      assignment.set(t, min);
    }

    for (const t of tests) {
      if (assignment.get(t) !== config.shard.current - 1)
        testRun.exclude(t);
    }
  }

  onBegin(config: FullConfig) {
    this._config = config;
  }

  onTestEnd(test: TestCase, result: TestResult) {
    this._durations.set(test.id, result.duration);
  }

  onEnd() {
    // merge-reports has no shard → persist timings from the full corpus.
    if (this._config.shard || !process.env.TIMINGS_FILE)
      return;
    fs.writeFileSync(
      process.env.TIMINGS_FILE,
      JSON.stringify(Object.fromEntries(this._durations)),
    );
  }
}
```

This reporter keeps timing data in a checked-in file in Git, and you can build workflows in GitHub Actions around updating this on a schedule. Since it is important that all shards receive the same timing data as input (otherwise you'll miss tests), I strongly recommend against using features like GitHub Actions Cache to store the timing data. It is not safe towards race conditions. I'm could imagine you can build an approach around GH artifacts and walking the git tree, though, that doesn't require checking in timing data.

Providers like Currents.dev or Flakiness.io can also implement this interface to balance distribution based on their collected history.

There's a limitation here around tests using the discouraged [serial mode](https://playwright.dev/docs/test-parallel#serial-mode), if you need that supported please open a separate issue.

**Comment by gpaciga** [comment]
> There's a limitation here around tests using the discouraged serial mode

@Skn0tt can you elaborate on what the limitation is? I will need support for it so need to know the nature of the problem. Unfortunately the same thing that causes us to have unbalanced tests in the first place (a few expensive test steps) means that serial mode is a necessary compromise for us.

**Comment by aslushnikov** [comment]
> Providers like Currents.dev or Flakiness.io can also implement this interface to balance distribution based on their collected history.

Yes, [`@flakiness/playwright`](https://github.com/flakiness/playwright) already uses this new API. Thank you @Skn0tt for working on this! WordPress's Gutenberg project has just adopted it: https://github.com/WordPress/gutenberg/pull/79479

We have a pretty solid implementation that supports both `serial` and `parallel` suites, as well as `fullyParallel` config option. We also designed an [advanced allocation algorithm](https://github.com/flakiness/playwright/blob/7c5f3c18829547bf47f1e25977c477d42b91f73e/src/sharding.ts#L48-L90). 

Btw, the reporter and its shard balancing algorithm can be used without Flakiness.io: https://blog.flakiness.io/posts/2026/playwright-test-shard-balancing/

**Comment by agoldis** [comment]
Thank you @aslushnikov @Skn0tt!

**Comment by parkuman** [comment]
Thanks all for getting this in!

**Comment by gpaciga** [comment]
> There's a limitation here around tests using the discouraged serial mode

For the record, the limitation here is that the reporter doesn't have information on which tests are serial or not, so it will divide a serial describe block across multiple shards. It's possible to implement a more complex version that uses `Suite._parallelMode`, but it's not part of the public API. I logged #42143 to add visibility on test groups (including serial tests) that reporters can leverage.

**Comment by vitalets** [comment]
Thanks for adding this possibility!

One concern I have is why this is being exposed through the reporter API? 
Rather than as a separate hook, similar to `globalSetup` and `globalTeardown`.

I assumed a reporter’s responsibility was only to report results, not to make decisions about how tests run. Could you share a bit more about why this approach was chosen?

**Comment by Skn0tt** [comment]
> Could you share a bit more about why this approach was chosen?

We contemplated this for long within the team, and ended up putting it into `Reporter` because, while it doesn't *perfectly* fit conceptually, it's the one thing in your config that the likely customers of this feature (flakiness.io et al) already have configured and can introduce support for with minimal changes needed. So the reason is pragmatism!
